clientapi/routing/login.go

```diff
@@ -53,40 +56,120 @@ func passwordLogin() flows {
 	return f
 }
 
+func ssoLogin() flows {
+	f := flows{}
+	s := flow{
+		Type: "m.login.sso",
+	}
+	f.Flows = append(f.Flows, s)
+	return f
+}
+
 // Login implements GET and POST /login
 func Login(
 	req *http.Request, accountDB accounts.Database, userAPI userapi.UserInternalAPI,
 	cfg *config.ClientAPI,
 ) util.JSONResponse {
 	if req.Method == http.MethodGet {
 		// TODO: support other forms of login other than password, depending on config options
+		flows := passwordLogin()
+		if cfg.CAS.Enabled {
+			flows.Flows = append(flows.Flows, ssoLogin().Flows...)
+		}
 		return util.JSONResponse{
 			Code: http.StatusOK,
-			JSON: passwordLogin(),
+			JSON: flows,
 		}
 	} else if req.Method == http.MethodPost {
-		typePassword := auth.LoginTypePassword{
-			GetAccountByPassword: accountDB.GetAccountByPassword,
-			Config:               cfg,
+		// TODO: is the the right way to read the body and re-add it?
+		body, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			// TODO: is this appropriate?
+			return util.JSONResponse{
+				Code: http.StatusMethodNotAllowed,
+				JSON: jsonerror.NotFound("Bad method"),
+			}
 		}
-		r := typePassword.Request()
-		resErr := httputil.UnmarshalJSONRequest(req, r)
-		if resErr != nil {
-			return *resErr
+		// add the body back to the request because ioutil.ReadAll consumes the body
+		req.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+
+		// marshall the body into an unstructured json map
+		var jsonBody map[string]interface{}
+		if err := json.Unmarshal([]byte(body), &jsonBody); err != nil {
```

Comment:
Use `json.NewDecoder(req.Body).Decode(&jsonBody)` instead of going to `body`.

Reply:
I tried this, but then I had to do `req.Body = ioutil.NopCloser(bytes.NewBuffer(body))` twice because the body is consumed when reading from it. Is that fine or is there something else that I'm missing?